bika/lims/browser/fields/aranalysesfield.py

```diff
@@ -99,158 +122,226 @@ def get(self, instance, **kwargs):
+        :returns: Analysis Service object or None
         """
-        bsc = get_tool('bika_setup_catalog')
-        brains = bsc(portal_type='AnalysisService')
-        return [proxy.getObject() for proxy in brains]
 
-    def _get_service_uid(self, item):
-        if api.is_uid(item):
-            return item
+        # Convert UIDs to objects
+        if api.is_uid(thing):
+            thing = api.get_object_by_uid(thing, None)
 
-        if not api.is_object(item):
-            logger.warn("Not an UID: {}".format(item))
+        # Bail out if the thing is not a valid object
+        if not api.is_object(thing):
+            logger.warn("'{}' is not a valid object!".format(repr(thing)))
             return None
 
-        obj = api.get_object(item)
+        # Ensure we have an object here and not a brain
+        obj = api.get_object(thing)
+
         if IAnalysisService.providedBy(obj):
-            return api.get_uid(obj)
+            return obj
 
-        if IAnalysis.providedBy(obj) and IRequestAnalysis.providedBy(obj):
-            return obj.getServiceUID()
+        if IAnalysis.providedBy(obj):
+            return obj.getAnalysisService()
 
         # An object, but neither an Analysis nor AnalysisService?
         # This should never happen.
         msg = "ARAnalysesField doesn't accept objects from {} type. " \
-              "The object will be dismissed."
-        logger.warn(msg.format(api.get_portal_type(obj)))
+            "The object will be dismissed.".format(api.get_portal_type(obj))
+        logger.warn(msg)
         return None
 
+    def _is_frozen(self, analysis):
+        """Check if the Analysis is in one of the frozen states
+
+        :param analysis: Analysis Brain/Object
+        :returns: True if the Analysis is in a frozen state
+        """
+        state = api.get_workflow_status_of(analysis, state_var='review_state')
+        return state in FROZEN_STATES
+
+    def _get_assigned_worksheets(self, analysis):
+        """Return the assigned worksheets of this Analysis
+
+        :param analysis: Analysis Brain/Object
+        :returns: Worksheet Backreferences
+        """
+        analysis = api.get_object(analysis)
+        return analysis.getBackReferences("WorksheetAnalysis")
+
+    def _is_assigned_to_worksheet(self, analysis):
+        """Check if the Analysis is assigned to a worksheet
+
+        :param analysis: Analysis Brain/Object
+        :returns: True if the Analysis is assigned to a WS
+        """
+        analysis = api.get_object(analysis)
+        state = api.get_workflow_status_of(
+            analysis, state_var='worksheetanalysis_review_state')
+        return state in ASSIGNED_STATES
+
+    def _update_interims(self, analysis, service):
+        """Update AR specifications
+
+        :param analysis: Analysis Object
+        :param service: Analysis Service Object
+        """
+        service_interims = service.getInterimFields()
+        analysis.setInterimFields(service_interims)
```

Comment:
The interim fields assigned to an AnalysisService are a copy of the calculation is associated to: when you create/edit an AnalysisService and assign a calculation, [the system gets the interim fields from the Calculation and assigns them to the field with the same name of the Analysis Service](https://github.com/senaite/senaite.core/blob/master/bika/lims/browser/js/bika.lims.analysisservice.js#L163-L253). Although done with javascript (much better if it was done in python), this is fine, but what happens if you change the Calculation after it was assigned previously assigned to an Analysis Service? The Analysis Service still has the interim fields from the calculation before it was changed. If I am not wrong, this is what the previous code was trying to overcome. I agree the previous approach was odd, but this scenario must be considered. I suggest to just update the interim fields from associated Analysis Services when a Calculation is updated.

Reply:
And if you can add a doctest with this use-case, then it would be awesome!

Reply:
The interim fields of the Analysis are only taken from the Analysis Service only, see here:

https://github.com/senaite/senaite.core/pull/593/files#diff-df7750f2a3730c33aaa0df4851641f7dR474

If the Interim Fields of the Calculation is changed, it reflects directly to each Analysis of the AS – not sure if dragons are sleeping here.

The code before did actually mesh the interims of both together in the Analysis, which is in my opinion wrong (if it would have set it, which it did not)

Reply:
I am not sure the interim fields from AS get updated automatically when they are changed on his Calculation counterpart. If yes, then I totally agree with `analysis.setInterimFields(service.getInterimFields())`. Give me a couple of hours to double-check.

Reply:
AS interims are currently not affected by the calculation Interims, see here in the doctest:

https://github.com/senaite/senaite.core/pull/593/files#diff-df7750f2a3730c33aaa0df4851641f7dR496

Therefore, I think putting them together on the Analysis doesn't make sense, but I've not analyzed the dependent code parts any further (If those merge the interims together or not)